OptiFineTweaker有时不应放在所有--tweakClass的末尾

* 启动器版本：3.2.143
* 操作系统：Windows 10 1909
* Java 版本：1.8.0_231
* 错误截图：无
* 游戏版本：1.14.4
* 问题描述：
我发现[这里](https://github.com/huanghongxun/HMCL/blob/bb2e74009e116f939d0035554664dc77fd6a35bd/HMCLCore/src/main/java/org/jackhuang/hmcl/download/MaintainTask.java#L89-L102)会导致OptiFineTweaker始终在所有--tweakClass的末尾，这在当使用[某些对--tweakClass参数有顺序要求的Tweaker](https://github.com/ImpactDevelopment/SimpleTweaker/blob/9feb1b1198fac58aecbd2bc871508e300bf8a721/src/main/java/io/github/impactdevelopment/simpletweaker/SimpleTweaker.java#L50-L52)时，会引发如下异常：
```
java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_231]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_231]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_231]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_231]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:159) [launchwrapper-of-2.1.jar:2.1]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:30) [launchwrapper-of-2.1.jar:2.1]
Caused by: joptsimple.MultipleArgumentsForOptionException: Found multiple arguments for option version, but you asked for only one
	at joptsimple.OptionSet.valueOf(OptionSet.java:179) ~[OptionSet.class:?]
	at net.minecraft.client.main.Main.a(SourceFile:164) ~[Main.class:?]
	at net.minecraft.client.main.Main.main(SourceFile:108) ~[Main.class:?]
	... 6 more
```

即使将下述启动配置json中的`OptiFineTweaker`修改为`OptiFineForgeTweaker`，HMCL也会将其替换回`OptiFineTweaker`
```json
{
  "id": "1.14.4-SimpleTweaker-1.2-OptiFine_HD_U_F5",
  "inheritsFrom": "1.14.4",
  "time": "1970-01-01T00:00:00+08:00",
  "releaseTime": "1970-01-01T00:00:00+08:00",
  "type": "release",
  "libraries": [
    {
      "name": "optifine:OptiFine:1.14.4_HD_U_F5"
    },
    {
      "name": "optifine:launchwrapper-of:2.1"
    },
    {
      "name": "org.ow2.asm:asm:7.3.1",
      "url": "https://repo1.maven.org/maven2/"
    },
    {
      "name": "org.ow2.asm:asm-analysis:7.3.1",
      "url": "https://repo1.maven.org/maven2/"
    },
    {
      "name": "org.ow2.asm:asm-commons:7.3.1",
      "url": "https://repo1.maven.org/maven2/"
    },
    {
      "name": "org.ow2.asm:asm-tree:7.3.1",
      "url": "https://repo1.maven.org/maven2/"
    },
    {
      "name": "org.ow2.asm:asm-util:7.3.1",
      "url": "https://repo1.maven.org/maven2/"
    },
    {
      "name": "com.github.ImpactDevelopment:SimpleTweaker:1.2",
      "url": "https://impactdevelopment.github.io/maven/"
    }
  ],
  "mainClass": "net.minecraft.launchwrapper.Launch",
  "minimumLauncherVersion": "21",
  "arguments": {
    "game": [
      "--tweakClass",
      "optifine.OptiFineTweaker",
      "--tweakClass",
      "io.github.impactdevelopment.simpletweaker.SimpleTweaker"
    ]
  }
}
```

--- a/HMCLCore/src/main/java/org/jackhuang/hmcl/download/LibraryAnalyzer.java
+++ b/HMCLCore/src/main/java/org/jackhuang/hmcl/download/LibraryAnalyzer.java
@@ -237,4 +237,15 @@ public final class LibraryAnalyzer implements Iterable<LibraryAnalyzer.LibraryMa
     public static final String LAUNCH_WRAPPER_MAIN = "net.minecraft.launchwrapper.Launch";
     public static final String MOD_LAUNCHER_MAIN = "cpw.mods.modlauncher.Launcher";
     public static final String BOOTSTRAP_LAUNCHER_MAIN = "cpw.mods.bootstraplauncher.BootstrapLauncher";
+
+    public static final String[] FORGE_TWEAKERS = new String[] {
+        "net.minecraftforge.legacy._1_5_2.LibraryFixerTweaker", // 1.5.2
+        "cpw.mods.fml.common.launcher.FMLTweaker", // 1.6.1 ~ 1.7.10
+        "net.minecraftforge.fml.common.launcher.FMLTweaker" // 1.8 ~ 1.12.2
+    };
+    public static final String[] OPTIFINE_TWEAKERS = new String[] {
+        "optifine.OptiFineTweaker",
+        "optifine.OptiFineForgeTweaker"
+    };
+    public static final String LITELOADER_TWEAKER = "com.mumfrey.liteloader.launch.LiteLoaderTweaker";
 }

--- a/HMCLCore/src/main/java/org/jackhuang/hmcl/download/MaintainTask.java
+++ b/HMCLCore/src/main/java/org/jackhuang/hmcl/download/MaintainTask.java
@@ -102,34 +102,42 @@ public class MaintainTask extends Task<Version> {
         VersionLibraryBuilder builder = new VersionLibraryBuilder(version);
         String mainClass = null;
 
-        if (!libraryAnalyzer.has(FORGE)) {
-            builder.removeTweakClass("forge");
-        }
-
         // Installing Forge will override the Minecraft arguments in json, so LiteLoader and OptiFine Tweaker are being re-added.
-
         if (libraryAnalyzer.has(LITELOADER) && !libraryAnalyzer.hasModLauncher()) {
-            builder.replaceTweakClass("liteloader", "com.mumfrey.liteloader.launch.LiteLoaderTweaker", !reorderTweakClass);
+            builder.replaceTweakClass(LibraryAnalyzer.LITELOADER_TWEAKER, LibraryAnalyzer.LITELOADER_TWEAKER, !reorderTweakClass, reorderTweakClass);
         } else {
-            builder.removeTweakClass("liteloader");
+            builder.removeTweakClass(LibraryAnalyzer.LITELOADER_TWEAKER);
         }
 
         if (libraryAnalyzer.has(OPTIFINE)) {
-            if (!libraryAnalyzer.has(LITELOADER) && !libraryAnalyzer.has(FORGE)) {
-                builder.replaceTweakClass("optifine", "optifine.OptiFineTweaker", !reorderTweakClass);
+            if (!libraryAnalyzer.has(LITELOADER) && !libraryAnalyzer.has(FORGE) && builder.hasTweakClass(LibraryAnalyzer.OPTIFINE_TWEAKERS[1])) {
+                builder.replaceTweakClass(LibraryAnalyzer.OPTIFINE_TWEAKERS[1], LibraryAnalyzer.OPTIFINE_TWEAKERS[0], !reorderTweakClass, reorderTweakClass);
             } else {
                 if (libraryAnalyzer.hasModLauncher()) {
                     // If ModLauncher installed, we use ModLauncher in place of LaunchWrapper.
-                    mainClass = "cpw.mods.modlauncher.Launcher";
-                    builder.replaceTweakClass("optifine", "optifine.OptiFineForgeTweaker", !reorderTweakClass);
+                    mainClass = LibraryAnalyzer.MOD_LAUNCHER_MAIN;
+                    for (String optiFineTweaker : LibraryAnalyzer.OPTIFINE_TWEAKERS) {
+                        builder.removeTweakClass(optiFineTweaker);
+                    }
                 } else {
                     // If forge or LiteLoader installed, OptiFine Forge Tweaker is needed.
-                    builder.replaceTweakClass("optifine", "optifine.OptiFineForgeTweaker", !reorderTweakClass);
+                    builder.replaceTweakClass(LibraryAnalyzer.OPTIFINE_TWEAKERS[0], LibraryAnalyzer.OPTIFINE_TWEAKERS[1], !reorderTweakClass, reorderTweakClass);
                 }
 
             }
         } else {
-            builder.removeTweakClass("optifine");
+            for (String optiFineTweaker : LibraryAnalyzer.OPTIFINE_TWEAKERS) {
+                builder.removeTweakClass(optiFineTweaker);
+            }
+        }
+
+        boolean hasForge = libraryAnalyzer.has(FORGE), hasModLauncher = libraryAnalyzer.hasModLauncher();
+        for (String forgeTweaker : LibraryAnalyzer.FORGE_TWEAKERS) {
+            if (!hasForge) {
+                builder.removeTweakClass(forgeTweaker);
+            } else if (!hasModLauncher && builder.hasTweakClass(forgeTweaker)) {
+                builder.replaceTweakClass(forgeTweaker, forgeTweaker, !reorderTweakClass, reorderTweakClass);
+            }
         }
 
         Version ret = builder.build();

--- a/HMCLCore/src/main/java/org/jackhuang/hmcl/game/VersionLibraryBuilder.java
+++ b/HMCLCore/src/main/java/org/jackhuang/hmcl/game/VersionLibraryBuilder.java
@@ -58,6 +58,10 @@ public final class VersionLibraryBuilder {
                 .setLibraries(libraries);
     }
 
+    public boolean hasTweakClass(String tweakClass) {
+        return useMcArgs && mcArgs.contains(tweakClass) || game.stream().anyMatch(arg -> arg.toString().equals(tweakClass));
+    }
+
     public void removeTweakClass(String target) {
         replaceTweakClass(target, null, false);
     }
@@ -94,6 +98,20 @@ public final class VersionLibraryBuilder {
      * @param inPlace if true, replace the tweak class in place, otherwise add the tweak class to the end of the argument list without replacement.
      */
     public void replaceTweakClass(String target, String replacement, boolean inPlace) {
+        replaceTweakClass(target, replacement, inPlace, false);
+    }
+
+    /**
+     * Replace existing tweak class.
+     * If the tweak class does not exist, the new tweak class will be added to argument list.
+     * If the tweak class appears more than one time, the tweak classes will be removed excluding the first one.
+     *
+     * @param target the tweak class to replace
+     * @param replacement the new tweak class to be replaced with, if null, remove the tweak class only
+     * @param inPlace if true, replace the tweak class in place, otherwise add the tweak class to the end of the argument list without replacement.
+     * @param reserve if true, add the tweak class to the start of the argument list.
+     */
+    public void replaceTweakClass(String target, String replacement, boolean inPlace, boolean reserve) {
         if (replacement == null && inPlace)
             throw new IllegalArgumentException("Replacement cannot be null in replace mode");
 
@@ -102,7 +120,7 @@ public final class VersionLibraryBuilder {
             for (int i = 0; i + 1 < mcArgs.size(); ++i) {
                 String arg0Str = mcArgs.get(i);
                 String arg1Str = mcArgs.get(i + 1);
-                if (arg0Str.equals("--tweakClass") && arg1Str.toLowerCase().contains(target)) {
+                if (arg0Str.equals("--tweakClass") && arg1Str.equals(target)) {
                     if (!replaced && inPlace) {
                         // for the first one, we replace the tweak class only.
                         mcArgs.set(i + 1, replacement);
@@ -124,7 +142,7 @@ public final class VersionLibraryBuilder {
                 // We need to preserve the tokens
                 String arg0Str = arg0.toString();
                 String arg1Str = arg1.toString();
-                if (arg0Str.equals("--tweakClass") && arg1Str.toLowerCase().contains(target)) {
+                if (arg0Str.equals("--tweakClass") && arg1Str.equals(target)) {
                     if (!replaced && inPlace) {
                         // for the first one, we replace the tweak class only.
                         game.set(i + 1, new StringArgument(replacement));
@@ -141,8 +159,18 @@ public final class VersionLibraryBuilder {
 
         // if the tweak class does not exist, add a new one to the end.
         if (!replaced && replacement != null) {
-            game.add(new StringArgument("--tweakClass"));
-            game.add(new StringArgument(replacement));
+            if (reserve) {
+                if (useMcArgs) {
+                    mcArgs.add(0, replacement);
+                    mcArgs.add(0, "--tweakClass");
+                } else {
+                    game.add(0, new StringArgument(replacement));
+                    game.add(0, new StringArgument("--tweakClass"));
+                }
+            } else {
+                game.add(new StringArgument("--tweakClass"));
+                game.add(new StringArgument(replacement));
+            }
         }
     }
 

